File: core-api/pkg/handler/schema/food.go
package schema

import (
	"time"

	"github.com/gin-gonic/gin"
)

type Food struct {
	Id                     string     `json:"id"`
	Name                   string     `json:"name"`
	ContainerWeightGram    float32    `json:"container_weight_gram"`
	ContainerMaxWeightGram float32    `json:"container_max_weight_gram"`
	GramPerMilliliter      *float32   `json:"gram_per_milliliter"`
	AccountId              string     `json:"account_id"`
	NfcUid                 string     `json:"nfc_uid"`
	RawWeightGram          float32    `json:"raw_weight_gram"`
	WeightGram             float32    `json:"weight_gram"`
	CreatedAt              *time.Time `json:"created_at,omitempty"`
	UpdatedAt              *time.Time `json:"updated_at,omitempty"`
}

type CreateFoodRequest struct {
	Name                   string   `json:"name"`
	NfcUid                 string   `json:"nfc_uid"`
	ContainerWeightGram    float32  `json:"container_weight_gram"`
	ContainerMaxWeightGram float32  `json:"container_max_weight_gram"`
	GramPerMilliliter      *float32 `json:"gram_per_milliliter"`
	Force                  bool     `json:"force"`
}

type UpdateFoodRequest struct {
	Id                     string   `json:"id"`
	Name                   string   `json:"name"`
	NfcUid                 string   `json:"nfc_uid"`
	ContainerWeightGram    float32  `json:"container_weight_gram"`
	ContainerMaxWeightGram float32  `json:"container_max_weight_gram"`
	GramPerMilliliter      *float32 `json:"gram_per_milliliter"`
	Force                  bool     `json:"force"`
}

type MyFoodsResponse struct {
	Foods          []*Food `json:"foods"`
	LowWeightFoods []*Food `json:"low_weight_foods"`
	UnusedFoods    []*Food `json:"unused_foods"`
}

type FoodController interface {
	CreateFood(c *gin.Context)
	UpdateFood(c *gin.Context)
	FindFoodByNfcUid(c *gin.Context)
	FindFoodById(c *gin.Context)
	FindByOwnFoods(c *gin.Context)
}

File: client/lib/schema/food.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'food.freezed.dart';
part 'food.g.dart';

@freezed
class Food with _$Food {
  const factory Food({
    required String id,
    required String name,
    @JsonKey(name: 'container_weight_gram') required double containerWeightGram,
    @JsonKey(name: 'container_max_weight_gram') required double containerMaxWeightGram,
    @JsonKey(name: 'gram_per_milliliter') double? gramPerMilliliter,
    @JsonKey(name: 'account_id') required String accountId,
    @JsonKey(name: 'nfc_uid') String? nfcUid,
    @JsonKey(name: 'raw_weight_gram') required double rawWeightGram,
    @JsonKey(name: 'weight_gram') required double weightGram,
    @JsonKey(name: 'created_at', fromJson: _fromJson, toJson: _toJson) DateTime? createdAt,
    @JsonKey(name: 'updated_at', fromJson: _fromJson, toJson: _toJson) DateTime? updatedAt,
  }) = _Food;

  factory Food.fromJson(Map<String, dynamic> json) => _$FoodFromJson(json);
}

@freezed
class CreateFoodRequest with _$CreateFoodRequest {
  const factory CreateFoodRequest({
    required String name,
    @JsonKey(name: 'nfc_uid') required String nfcUid,
    @JsonKey(name: 'container_weight_gram') required double containerWeightGram,
    @JsonKey(name: 'container_max_weight_gram') required double containerMaxWeightGram,
    @JsonKey(name: 'gram_per_milliliter') double? gramPerMilliliter,
    required bool force,
  }) = _CreateFoodRequest;

  factory CreateFoodRequest.fromJson(Map<String, dynamic> json) => _$CreateFoodRequestFromJson(json);
}

@freezed
class UpdateFoodRequest with _$UpdateFoodRequest {
  const factory UpdateFoodRequest({
    required String id,
    required String name,
    @JsonKey(name: 'nfc_uid') required String nfcUid,
    @JsonKey(name: 'container_weight_gram') required double containerWeightGram,
    @JsonKey(name: 'container_max_weight_gram') required double containerMaxWeightGram,
    @JsonKey(name: 'gram_per_milliliter') double? gramPerMilliliter,
    required bool force,
  }) = _UpdateFoodRequest;

  factory UpdateFoodRequest.fromJson(Map<String, dynamic> json) => _$UpdateFoodRequestFromJson(json);
}

@freezed
class MyFoodsResponse with _$MyFoodsResponse {
  const factory MyFoodsResponse({
    @JsonKey(name: 'foods') required List<Food> foods,
    @JsonKey(name: 'low_weight_foods') required List<Food> lowWeightFoods,
    @JsonKey(name: 'unused_foods') required List<Food> unusedFoods,
  }) = _MyFoodsResponse;

  factory MyFoodsResponse.fromJson(Map<String, dynamic> json) => _$MyFoodsResponseFromJson(json);
}

DateTime? _fromJson(String? dateString) => dateString != null ? DateTime.parse(dateString) : null;
String? _toJson(DateTime? dateTime) => dateTime?.toIso8601String();
